Select.dictionary element string wildcard

thomasstvr opened the pull request.

ChrisWRWX commented:
There's a function called wildcard_expansion in utils.py that's used for the column wildcards.

It would be good to reuse that here. It deals with the *, but has a bunch more features like regex: <pattern>, -value for not and value? for optional.

thomasstvr commented:
Ok, I _think_ this may have been what you'd had in mind.

Note that the output for both regex and wildcards that match only a single element is a dictionary.

thomasstvr commented:
Dropped the check for * or regex and added a recursive run to handle wildcard expansion.

Regex and wildcard elements that are passed as strings now output a string when only a single element is selected, but a dictionary when multiple elements are selected. Rewrote tests to reflect this, but it should still be inline with how this wrangle has worked in the past.

ebhills reviewed (changes requested):
Let's stick with a dictionary output even if 1 or 0 matches whne input is * or regex.

thomasstvr commented on wrangles/select.py:
```diff
@@ -170,13 +172,31 @@ def _get_value(value):
+
+        # If key is now a list after wildcard expansion, send back through dict_element
+        if isinstance(key, list):
+            results = dict_element(input, key, default)
```
Looking at these two lines of code alone, it does appear to have this issue, but doesn't the if statement on line 120 cover this?

```python
    if isinstance(key, list):
        key = dict(
            _itertools.chain.from_iterable(
                [
                    x.items() if isinstance(x, dict)
                    else {x: x}.items()
                    for x in key
                ]
            )
        )
```
